rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -145,12 +183,126 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
 
         maybe_reshuffle_events(out_event=out_event, in_event=in_event)
 
+    def _maybe_decode_v3_deposit_or_withdrawal(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[list[ActionItem]],  # pylint: disable=unused-argument
+            event_action_type: Literal['addition', 'removal'],
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        """
+        This method decodes a Uniswap V3 LP liquidity increase or decrease.
+
+        Examples of such transactions are:
+        https://etherscan.io/tx/0x6bf3588f669a784adf5def3c0db149b0cdbcca775e472bb35f00acedee263c4c (deposit)
+        https://etherscan.io/tx/0x76c312fe1c8604de5175c37dcbbb99cc8699336f3e4840e9e29e3383970f6c6d (withdrawal)
+        """  # noqa: E501
+        new_action_items = []
+        liquidity_pool_id = hex_or_bytes_to_int(tx_log.topics[1])
+        amount0_raw = hex_or_bytes_to_int(tx_log.data[32:64])
+        amount1_raw = hex_or_bytes_to_int(tx_log.data[64:96])
+
+        if event_action_type == 'addition':
+            notes = 'Add {} {} of liquidity to Uniswap V3 LP position {}'
+            from_event_type = (HistoryEventType.SPEND, HistoryEventSubType.NONE)
+            to_event_type = (HistoryEventType.DEPOSIT, HistoryEventSubType.DEPOSIT_ASSET)
+        else:
+            notes = 'Collect {} {} of liquidity from Uniswap V3 LP position {}'
+            from_event_type = (HistoryEventType.RECEIVE, HistoryEventSubType.NONE)
+            to_event_type = (HistoryEventType.WITHDRAWAL, HistoryEventSubType.REMOVE_ASSET)
+
+        liquidity_pool_position_info = self.ethereum_inquirer.contracts.contract('UNISWAP_V3_NFT_MANAGER').call(  # noqa: E501
+            node_inquirer=self.ethereum_inquirer,
+            method_name='positions',
+            arguments=[liquidity_pool_id],
+        )
+        # 2 -> first token in the pair
+        # 3 -> second token in the pair
+        for token, amount in zip(liquidity_pool_position_info[2:4], (amount0_raw, amount1_raw)):
+            token_with_data: 'CryptoAsset' = get_or_create_evm_token(
```

Comment:
> can you state that this is a for loop that executes only twice?

Won't that be excessive commenting? Given that the indexes(2:4) and other list(length of 2) give the number of execution of the loop.

Reply:
> The problem with this for loop is that you are iterating twice the list of decoded events that can be huge. I'm not sure if it is over engineering to avoid that or should be a required change

I do share your view on how looping through what might be a large list twice might be an expensive operation. However, the only alternative that comes to mind would involve code repetition. If you do have a different view on how this can be approached, I'm open to such.

Reply:
I also think this needs refactoring here. I get that it looks and reads nice, but we should not sacrifice speed for readability to such an extent. I was gonna try and come up with a nice alternative but it's more tricky than what I can do at 01:30 at night.

So please think on this a bit and try to find a way to have this read as nicely as this does but also not iterate the events twice.